org.eclipse.wildwebdeveloper.xml/src/org/eclipse/wildwebdeveloper/xml/XMLLSClassPathExtensionRegistry.java

```diff
@@ -0,0 +1,76 @@
+/*******************************************************************************
+ * Copyright (c) 2019 Red Hat Inc. and others.
+ *
+ * This program and the accompanying materials are made
+ * available under the terms of the Eclipse Public License 2.0
+ * which is available at https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *   Andrew Obuchowicz (Red Hat Inc.) - initial implementation
+ *******************************************************************************/
+package org.eclipse.wildwebdeveloper.xml;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.Set;
+
+import org.eclipse.core.runtime.IConfigurationElement;
+import org.eclipse.core.runtime.IStatus;
+import org.eclipse.core.runtime.Platform;
+import org.eclipse.core.runtime.Status;
+
+public class XMLLSClassPathExtensionRegistry {
+	private static final String EXTENSION_POINT_ID = Activator.PLUGIN_ID + ".xmllsClasspathExtensionProvider"; //$NON-NLS-1$
+	private Map<IConfigurationElement, XMLLSClasspathExtensionProvider> extensions = new HashMap<>();
+	private boolean outOfSync = true;
+
+	public XMLLSClassPathExtensionRegistry() {
+		Platform.getExtensionRegistry().addRegistryChangeListener((event -> outOfSync = true), EXTENSION_POINT_ID);
+	}
+
+	public List<String> getXMLLSClassPathExtensions() {
+		if (this.outOfSync) {
+			sync();
+		}
+
+		List<String> classpathExtensions = new ArrayList<>();
+		this.extensions.entrySet().stream()
+				.map(Entry<IConfigurationElement, XMLLSClasspathExtensionProvider>::getValue)
+				.map(XMLLSClasspathExtensionProvider::get)
+				.forEach(list -> list.forEach(jar -> classpathExtensions.add(jar.getAbsolutePath())));
+
+		return classpathExtensions;
+	}
+
+	private void sync() {
+		Set<IConfigurationElement> toRemoveExtensions = new HashSet<>(this.extensions.keySet());
```

Comment:
I'll move this logic (line 52 to 69) to the `getXMLLSClassPathExtensions` function? (As we agreed there's no need for the sync pattern here)

Reply:
Both works. I let you decide ;)